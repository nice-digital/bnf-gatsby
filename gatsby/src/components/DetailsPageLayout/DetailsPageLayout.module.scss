@use '@nice-digital/nds-core/scss/core';

.body {
  > :first-child {
    margin-top: 0;
  }
}

.toggleButton {
  appearance: none;
  background: core.$nds-colour-nice-grey-3;
  border: 1px solid core.$nds-colour-border;
  color: core.$nds-colour-link;
  cursor: pointer;
  display: block;
  line-height: 1;
  margin: 0;
  /* stylelint-disable indentation, function-parentheses-space-inside */
  padding: core.rem(
    core.$nds-spacing-medium core.$nds-spacing-large core.$nds-spacing-medium
      core.$nds-spacing-small
  );
  position: relative;
  text-align: left;
  text-decoration: none;
  width: 100%;

  &:visited {
    color: core.$nds-colour-link;
  }

  &:focus {
    @include core.nds-default-focus-style;
    z-index: 1;
  }

  /* stylelint-disable selector-class-pattern */
  + :global(.stacked-nav) {
    display: none;

    // Non-JS fallback
    &:target {
      display: block;
    }
  }

  &[aria-expanded='false'] + :global(.stacked-nav) {
    display: none;
  }

  &[aria-expanded='true'] {
    border-bottom: 0;

    .icon {
      transform: translateY(-50%) rotate(180deg);
    }

    + :global(.stacked-nav) {
      display: block;
    }
  }

  @include core.mq($from: sm) {
    display: none;
  }
}

@include core.mq($from: sm) {
  :global(.stacked-nav) {
    display: block !important;
  }
}
