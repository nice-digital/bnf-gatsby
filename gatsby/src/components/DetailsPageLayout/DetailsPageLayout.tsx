import { Link } from "gatsby";
import React, { useMemo, type ReactNode, type ElementType } from "react";
import striptags from "striptags";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";

import { Layout } from "@/components/Layout/Layout";
import { SEO } from "@/components/SEO/SEO";
import { useSiteMetadata } from "@/hooks/useSiteMetadata";

type DetailsPageLayoutProps = {
	/** The ancestors from the parent page e.g. ["About"] */
	parentTitleParts?: string[];
	titleHtml: string;
	menu: ElementType;
	children: ReactNode;
	parentBreadcrumbs?: {
		href: string;
		text: string;
	}[];
	metaDescription?: string;
};

/**
 * A 3 column layout with menu, main content and 'on this page'
 */
export const DetailsPageLayout: React.FC<DetailsPageLayoutProps> = ({
	children,
	parentTitleParts = [],
	titleHtml,
	menu: Menu,
	parentBreadcrumbs = [],
	metaDescription,
}) => {
	const { siteTitleShort } = useSiteMetadata(),
		titleNoHtml = striptags(titleHtml),
		breadcrumbElements = useMemo(
			() => [
				<Breadcrumb key="NICE" to="https://www.nice.org.uk/">
					NICE
				</Breadcrumb>,
				<Breadcrumb key="Home" to="/" elementType={Link}>
					{siteTitleShort}
				</Breadcrumb>,
				...parentBreadcrumbs.map(({ href, text }) => (
					<Breadcrumb key={href} to={href} elementType={Link}>
						{text}
					</Breadcrumb>
				)),
				<Breadcrumb key="Current page">{titleNoHtml}</Breadcrumb>,
			],
			[siteTitleShort, titleNoHtml, parentBreadcrumbs]
		);

	return (
		<Layout>
			<SEO
				title={[titleNoHtml, ...parentTitleParts].filter(Boolean).join(" | ")}
				description={metaDescription}
			/>

			<Breadcrumbs>{breadcrumbElements}</Breadcrumbs>

			<PageHeader
				id="content-start"
				heading={<span dangerouslySetInnerHTML={{ __html: titleHtml }} />}
			/>

			<Grid gutter="loose" data-testid="body">
				<GridItem cols={12} md={4} lg={3} className="hide-print">
					<Menu />
				</GridItem>
				<GridItem cols={12} md={8} lg={9}>
					<Grid reverse gutter="loose">
						<GridItem cols={12} lg={3}>
							<nav aria-labelledby="on-this-page" className="hide-print">
								<h2 id="on-this-page" className="h4">
									On this page
								</h2>
								<ol aria-label="Jump links to sections on this page">
									<li>TODO</li>
									{/* {sections.map((section, i) => (
										<li key={section.slug}>
											<a href={`#${section.slug}`}>
												<span
													dangerouslySetInnerHTML={{ __html: section.title }}
												/>
											</a>
										</li>
									))} */}
								</ol>
							</nav>
						</GridItem>
						<GridItem cols={12} lg={9}>
							{children}
						</GridItem>
					</Grid>
				</GridItem>
			</Grid>
		</Layout>
	);
};
