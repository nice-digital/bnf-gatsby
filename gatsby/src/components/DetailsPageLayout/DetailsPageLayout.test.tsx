/* eslint-disable testing-library/no-render-in-setup */
/* eslint-disable testing-library/no-node-access */
import { render, screen, waitFor, within } from "@testing-library/react";

import { DetailsPageLayout } from "./DetailsPageLayout";

describe("DetailsPageLayout", () => {
	describe("SEO", () => {
		beforeEach(() => {
			render(
				<DetailsPageLayout
					menu={() => null}
					titleHtml="FAQs for BNF <em>for Children</em> — general"
					parentTitleParts={["Parent", "Grandparent"]}
					metaDescription="A test meta"
				>
					test
				</DetailsPageLayout>
			);
		});

		it("should strip tags from titleHtml prop in page <title> tag", async () => {
			await waitFor(() => {
				expect(document.title).toStartWith(
					"FAQs for BNF for Children — general"
				);
			});
		});

		it("should append title parent title parts to page title", async () => {
			await waitFor(() => {
				expect(document.title).toStartWith(
					"FAQs for BNF for Children — general | Parent | Grandparent"
				);
			});
		});

		it("should set page meta description", async () => {
			await waitFor(() => {
				expect(
					document
						.querySelector("meta[name='description']")
						?.getAttribute("content")
				).toEqual("A test meta");
			});
		});
	});

	describe("Breadcrumbs", () => {
		beforeEach(() => {
			render(
				<DetailsPageLayout
					menu={() => null}
					titleHtml="Some title"
					parentBreadcrumbs={[{ href: "/about/", text: "About" }]}
				>
					test
				</DetailsPageLayout>
			);
		});

		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render given parent breadcrumbs", () => {
			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const breadcrumb = within(breadcrumbNav).getByRole("link", {
				name: "About",
			});

			expect(breadcrumb).toHaveAttribute("href", "/about/");
		});

		it("should render current page breadcrumb without link", () => {
			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb = within(breadcrumbNav).getByText("Some title");
			expect(currentPageCrumb).toBeInTheDocument();
			expect(currentPageCrumb.tagName).toBe("SPAN");
		});
	});

	describe("Page header", () => {
		it("should add content start skip link target id to page header", () => {
			render(
				<DetailsPageLayout menu={() => null} titleHtml="Anything">
					test
				</DetailsPageLayout>
			);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1.parentElement).toHaveProperty("id", "content-start");
		});

		it("should render heading 1 with current page title", () => {
			render(
				<DetailsPageLayout menu={() => null} titleHtml="Some <b>title</b>">
					test
				</DetailsPageLayout>
			);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1).toHaveTextContent("Some title");
			expect(heading1).toHaveProperty(
				"innerHTML",
				"<span>Some <b>title</b></span>"
			);
		});
	});

	describe("Body", () => {
		it("should match snapshot for body", () => {
			render(
				<DetailsPageLayout menu={() => <>Menu</>} titleHtml="Anything">
					Body content
				</DetailsPageLayout>
			);
			expect(screen.getByTestId("body")).toMatchSnapshot();
		});
	});
});
