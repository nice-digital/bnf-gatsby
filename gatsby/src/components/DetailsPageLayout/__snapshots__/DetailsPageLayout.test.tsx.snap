// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPageLayout Body should match snapshot for body 1`] = `
<div
  class="grid grid--loose"
  data-testid="body"
>
  <div
    class="hide-print"
    data-g="12 md:4 lg:3"
  >
    Menu
  </div>
  <div
    data-g="12 md:8 lg:9"
  >
    <div
      class="grid grid--rev grid--loose"
    >
      <div
        data-g="12 lg:3"
      >
        <nav
          aria-labelledby="on-this-page"
          class="hide-print"
        >
          <h2
            class="h4"
            id="on-this-page"
          >
            On this page
          </h2>
          <ol
            aria-label="Jump links to sections on this page"
          >
            <li>
              TODO
            </li>
          </ol>
        </nav>
      </div>
      <div
        data-g="12 lg:9"
      >
        Body content
      </div>
    </div>
  </div>
</div>
`;
